samcli/local/lambdafn/runtime.py

```diff
@@ -161,28 +242,278 @@ def _get_code_dir(self, code_path):
+        if container and container.is_created():
+            LOG.info("Reuse the created warm container for Lambda function '%s'", function_config.name)
+            return container
+
+        # debug_context should be used only if the function name is the one defined
+        # in debug-function option
+        if debug_context and debug_context.debug_function != function_config.name:
+            LOG.debug(
+                "Disable the debugging for Lambda Function %s, as the passed debug function is %s",
+                function_config.name,
+                debug_context.debug_function,
+            )
+            debug_context = None
+
+        container = super().create(function_config, debug_context, None)
+        self._containers[function_config.name] = container
+        self._add_function_to_observer(function_config)
+        return container
+
+    def _on_invoke_done(self, container):
+        """
+        Cleanup the created resources, just before the invoke function ends.
+        In warm containers, the running containers will be closed just before the end of te command execution,
+        so no action is done here
 
-            elif os.path.isfile(code_path) and code_path.endswith(self.SUPPORTED_ARCHIVE_EXTENSIONS):
+        Parameters
+        ----------
+        container: Container
+           The current running container
+        """
+
+    def _add_function_to_observer(self, function_config):
+        """
+        observe the function code path, so the function container be reload if its code got changed.
 
-                decompressed_dir = _unzip_file(code_path)
-                yield decompressed_dir
+        Parameters
+        ----------
+        function_config: FunctionConfig
+            Configuration of the function to create a new Container for it.
+        """
+        with self._lock:
+            code_paths = [function_config.code_abs_path]
+            if function_config.layers:
+                code_paths += [layer.codeuri for layer in function_config.layers]
+            # get all sub directories to be observed, so if the any of the sub directories got changed
+            # we can map it to the correct lambda function
+            code_paths_sub_folders = []
+            for code_path in code_paths:
+                code_paths_sub_folders += [code_path]
+                if not os.path.isfile(code_path):
+                    code_paths_sub_folders += [sub_dir[0] for sub_dir in os.walk(code_path)]
+
+            # observe all the code directories && files
+            for code_path in code_paths_sub_folders:
+                if os.path.isfile(code_path):
+                    checksum = file_checksum(code_path)
+                else:
+                    # observe the build directory instead of the lambda function directory in case of runtime languages
+                    # that needs compilation
+                    if ".aws-sam" in code_path:
```

Comment:
Why are we checking `.aws-sam` directly here? Why is it not just the code_path (CodeUri/Code) set in the template?